File: Clients/QSC_AI_Generated/metrics_client.lua
--[[
AV-System-Metrics Q-SYS Lua client.

Copy this file into a Q-SYS Lua Module named "metrics_client", then require it
from a Control Script:

    local Metrics = require("metrics_client")

    metrics = Metrics.new({
        processor_name = "boardroom-qsys-core",
        uri_type = "aws_lambda",
        uri = "https://example.lambda-url.us-west-1.on.aws/",
        bearer_token = "change-me-long-random-token"
    })

    metrics:trace("System Initialized")
]]

local rapidjson = require("rapidjson")

local Metrics = {}
Metrics.__index = Metrics
Metrics.__version__ = "2.1.2"

local MAX_BATCH_SIZE = 25

local function default_logger(message, level)
    print(string.format("%s: %s", string.upper(tostring(level or "info")), tostring(message)))
end

local function is_blank(value)
    return value == nil or tostring(value) == ""
end

local function current_seconds()
    if Timer ~= nil and Timer.Now ~= nil then
        local ok, value = pcall(Timer.Now)
        if ok and type(value) == "number" then
            return value
        end
    end

    if os ~= nil and os.time ~= nil then
        return os.time()
    end

    return 0
end

local function utc_timestamp()
    if os ~= nil and os.date ~= nil then
        local ok, value = pcall(os.date, "!%Y-%m-%dT%H:%M:%SZ")
        if ok and type(value) == "string" then
            return value
        end

        ok, value = pcall(os.date, "%Y-%m-%dT%H:%M:%SZ")
        if ok and type(value) == "string" then
            return value
        end
    end

    return "1970-01-01T00:00:00Z"
end

local function get_setting(settings, snake_name, camel_name, default_value)
    local value = settings[snake_name]
    if value == nil and camel_name ~= nil then
        value = settings[camel_name]
    end
    if value == nil then
        return default_value
    end
    return value
end

local function get_number(value, default_value)
    local number_value = tonumber(value)
    if number_value == nil then
        return default_value
    end
    return number_value
end

function Metrics.new(logger_or_settings, processor_name, uri_type, uri, bearer_token,
                     batch_size, flush_interval, max_cache_size,
                     failure_drop_message_threshold, failure_drop_time_threshold,
                     request_timeout)
    local self = setmetatable({}, Metrics)

    if type(logger_or_settings) == "table" then
        local settings = logger_or_settings
        self.logger = get_setting(settings, "logger", nil, default_logger)
        self.processor_name = get_setting(settings, "processor_name", "processorName", nil)
        if self.processor_name == nil then
            self.processor_name = get_setting(settings, "client_name", "clientName", settings.clientname)
        end
        self.uri_type = string.lower(tostring(get_setting(settings, "uri_type", "uriType", "")))
        self.uri = get_setting(settings, "uri", nil, nil)
        self.bearer_token = get_setting(settings, "bearer_token", "bearerToken", nil)
        self.batch_size = get_number(get_setting(settings, "batch_size", "batchSize", 20), 20)
        self.flush_interval = get_number(get_setting(settings, "flush_interval", "flushInterval", 10), 10)
        self.max_cache_size = get_number(get_setting(settings, "max_cache_size", "maxCacheSize", 250), 250)
        self.failure_drop_message_threshold = get_number(get_setting(settings, "failure_drop_message_threshold", "failureDropMessageThreshold", 100), 100)
        self.failure_drop_time_threshold = get_number(get_setting(settings, "failure_drop_time_threshold", "failureDropTimeThreshold", 300), 300)
        self.request_timeout = get_number(get_setting(settings, "request_timeout", "requestTimeout", 10), 10)
    else
        self.logger = logger_or_settings or default_logger
        self.processor_name = processor_name
        self.uri_type = string.lower(tostring(uri_type or ""))
        self.uri = uri
        self.bearer_token = bearer_token
        self.batch_size = get_number(batch_size or 20, 20)
        self.flush_interval = get_number(flush_interval or 10, 10)
        self.max_cache_size = get_number(max_cache_size or 250, 250)
        self.failure_drop_message_threshold = get_number(failure_drop_message_threshold or 100, 100)
        self.failure_drop_time_threshold = get_number(failure_drop_time_threshold or 300, 300)
        self.request_timeout = get_number(request_timeout or 10, 10)
    end

    self._metric_cache = {}
    self._flush_scheduled = false
    self._sending = false
    self._failed_message_count = 0
    self._first_failure_time = nil

    self:_validate_settings()
    self:_log("Metrics settings validated successfully", "info")

    return self
end

Metrics.New = Metrics.new

function Metrics:_validate_settings()
    if Timer == nil or Timer.CallAfter == nil then
        error("Q-SYS Timer.CallAfter is required")
    end

    if HttpClient == nil or HttpClient.Upload == nil then
        error("Q-SYS HttpClient.Upload is required")
    end

    if self.uri_type ~= "aws_lambda" and self.uri_type ~= "self-hosted" and self.uri_type ~= "aws_api_gateway" then
        error("uri_type must be one of: aws_lambda, self-hosted, aws_api_gateway")
    end

    if self.uri_type == "aws_api_gateway" then
        error("aws_api_gateway is not implemented yet for AV-System-Metrics")
    end

    if is_blank(self.processor_name) then
        error("processor_name can not be empty")
    end

    if is_blank(self.uri) then
        error("uri can not be empty")
    end

    if is_blank(self.bearer_token) then
        error("bearer_token can not be empty")
    end

    if self.batch_size < 1 or self.batch_size > MAX_BATCH_SIZE then
        error("batch_size must be between 1 and 25")
    end

    if self.flush_interval <= 0 then
        error("flush_interval must be greater than 0")
    end

    if self.max_cache_size < self.batch_size then
        error("max_cache_size must be greater than or equal to batch_size")
    end

    if self.failure_drop_message_threshold < 1 then
        error("failure_drop_message_threshold must be at least 1")
    end

    if self.failure_drop_time_threshold <= 0 then
        error("failure_drop_time_threshold must be greater than 0")
    end

    if self.request_timeout <= 0 then
        error("request_timeout must be greater than 0")
    end
end

function Metrics:_log(message, level)
    if type(self.logger) == "function" then
        pcall(self.logger, tostring(message), level or "info")
    else
        default_logger(message, level)
    end
end

function Metrics:_can_drop_oldest_metrics()
    if self._failed_message_count >= self.failure_drop_message_threshold then
        return true
    end

    if self._first_failure_time ~= nil then
        return (current_seconds() - self._first_failure_time) >= self.failure_drop_time_threshold
    end

    return false
end

function Metrics:_trim_cache_if_allowed()
    if #self._metric_cache <= self.max_cache_size then
        return
    end

    if not self:_can_drop_oldest_metrics() then
        self:_log(
            string.format(
                "Metric cache size is %d, max is %d, but dropping is not allowed yet.",
                #self._metric_cache,
                self.max_cache_size
            ),
            "warning"
        )
        return
    end

    local drop_count = #self._metric_cache - self.max_cache_size
    for _ = 1, drop_count do
        table.remove(self._metric_cache, 1)
    end

    self:_log(
        string.format(
            "Dropping %d oldest cached metrics after send failures. Failed message count: %d",
            drop_count,
            self._failed_message_count
        ),
        "warning"
    )
end

function Metrics:_cache_metric(action, metric_name)
    table.insert(self._metric_cache, {
        clientname = tostring(self.processor_name),
        timestamp = utc_timestamp(),
        metric = tostring(metric_name or ""),
        action = tostring(action or "")
    })

    self:_trim_cache_if_allowed()

    if #self._metric_cache >= self.batch_size then
        self:flush()
    else
        self:_schedule_flush()
    end
end

function Metrics:_schedule_flush()
    if self._flush_scheduled then
        return
    end

    self._flush_scheduled = true

    Timer.CallAfter(function()
        self._flush_scheduled = false
        self:flush()
    end, self.flush_interval)
end

function Metrics:_take_batch()
    local count = math.min(self.batch_size, #self._metric_cache)
    local batch = {}

    for i = 1, count do
        batch[i] = self._metric_cache[i]
    end

    for _ = 1, count do
        table.remove(self._metric_cache, 1)
    end

    return batch
end

function Metrics:flush()
    if self._sending then
        self:_schedule_flush()
        return
    end

    if #self._metric_cache == 0 then
        return
    end

    self._sending = true
    self:_send_batch(self:_take_batch())
end

Metrics.Flush = Metrics.flush

function Metrics:_send_batch(messages)
    local payload, encode_error = rapidjson.encode({ messages = messages })
    if payload == nil then
        self:_log("Could not encode metrics payload: " .. tostring(encode_error), "error")
        self:_record_send_failure(messages)
        self:_finish_send()
        return
    end

    local ok, upload_error = pcall(function()
        HttpClient.Upload {
            Url = self.uri,
            Method = "POST",
            Data = payload,
            Headers = {
                ["Content-Type"] = "application/json",
                ["Authorization"] = "Bearer " .. tostring(self.bearer_token)
            },
            Timeout = self.request_timeout,
            EventHandler = function(_, code, data, err)
                self:_handle_response(code, data, err, messages)
            end
        }
    end)

    if not ok then
        self:_log(upload_error, "error")
        self:_record_send_failure(messages)
        self:_finish_send()
    end
end

function Metrics:_handle_response(code, data, err, messages)
    local status_code = tonumber(code) or 0

    if status_code >= 200 and status_code < 300 then
        self:_log("Server response status: " .. tostring(status_code), "info")
        if data ~= nil and tostring(data) ~= "" then
            self:_log(data, "info")
        end

        local ack_ok, ack_error = self:_validate_server_ack(data, #messages)
        if ack_ok then
            self:_reset_failure_state()
        else
            self:_log(
                string.format(
                    "Server acknowledgement validation failed: %s. Will retry %d metrics.",
                    tostring(ack_error),
                    #messages
                ),
                "error"
            )
            self:_record_send_failure(messages)
        end

        self:_finish_send()
        return
    end

    if status_code == 429 then
        self:_log(
            string.format("Server rate limited request (429). Will retry %d metrics.", #messages),
            "warning"
        )
        self:_record_send_failure(messages)
    elseif status_code >= 400 and status_code < 500 then
        self:_log(
            string.format("Permanent client error %d. Dropping %d metrics.", status_code, #messages),
            "error"
        )
        if data ~= nil and tostring(data) ~= "" then
            self:_log(data, "error")
        end
    elseif status_code >= 500 then
        self:_log(
            string.format("Server error %d. Will retry %d metrics.", status_code, #messages),
            "error"
        )
        self:_record_send_failure(messages)
    else
        self:_log(
            string.format("HTTP request failed. Code=%s Error=%s. Will retry %d metrics.", tostring(code), tostring(err), #messages),
            "error"
        )
        self:_record_send_failure(messages)
    end

    self:_finish_send()
end

function Metrics:_validate_server_ack(data, expected_count)
    if data == nil or tostring(data) == "" then
        return false, "empty response body"
    end

    local decode_ok, ack, decode_error = pcall(rapidjson.decode, tostring(data))
    if not decode_ok then
        return false, "invalid JSON response: " .. tostring(ack)
    end

    if ack == nil then
        return false, "invalid JSON response: " .. tostring(decode_error)
    end

    if type(ack) ~= "table" then
        return false, "response JSON was not an object"
    end

    if ack.ok ~= true then
        return false, "response ok was not true"
    end

    if type(ack.count) ~= "number" then
        return false, "response count was not a number"
    end

    if ack.count ~= expected_count then
        return false, string.format(
            "response count %s did not match sent count %d",
            tostring(ack.count),
            expected_count
        )
    end

    return true, nil
end

function Metrics:_finish_send()
    self._sending = false

    if #self._metric_cache > 0 then
        self:_schedule_flush()
    end
end

function Metrics:_record_send_failure(messages)
    if self._first_failure_time == nil then
        self._first_failure_time = current_seconds()
    end

    self._failed_message_count = self._failed_message_count + #messages

    for i = #messages, 1, -1 do
        table.insert(self._metric_cache, 1, messages[i])
    end

    self:_trim_cache_if_allowed()
end

function Metrics:_reset_failure_state()
    self._failed_message_count = 0
    self._first_failure_time = nil
end

function Metrics:start(metric_name)
    self:_cache_metric("Started", metric_name)
end

Metrics.Start = Metrics.start

function Metrics:stop(metric_name)
    self:_cache_metric("Stopped", metric_name)
end

Metrics.Stop = Metrics.stop

function Metrics:trace(metric_name)
    self:_cache_metric("Trace", metric_name)
end

Metrics.Trace = Metrics.trace

function Metrics:custom(action, metric_name)
    self:_cache_metric(action, metric_name)
end

Metrics.Custom = Metrics.custom

return Metrics

File: Clients/Extron ECS/metrics_client.py
"""
Extron Processor Module for sending batched metric data.

You can also run this module on a workstation for testing.

Requirements: A server or serverless system for ingestion.
Find the ingest system code and documentation here: https://github.com/mefranklin6/AV-System-Metrics

Usage Example:

from metrics_client import Metrics

metrics = Metrics(
    logger=logger, # such as your instance of ProgramLog
    processor_name="my_processor",
    uri_type="aws_lambda",
    uri="https://myrandomlambdainstance.lambda-url.us-west-1.on.aws/",
    bearer_token="myrandombearertoken",
)

metrics.trace("Hello World!")
"""

from datetime import datetime, timezone
import json
import time
import urllib.error
import urllib.request

try:
    from extronlib.system import Wait  # type: ignore
except ImportError:  # Not running on Extron processor
    # Mock extronlib classes here so we can run on a workstation for testing

    from functools import wraps
    import threading

    class Wait:
        def __init__(self, delay_time: float = 0):
            self.delay_time = delay_time

        def __call__(self, func):
            @wraps(func)
            def wrapper(*args, **kwargs):
                return func(*args, **kwargs)

            timer = threading.Timer(self.delay_time, wrapper)
            timer.daemon = True
            timer.start()

            return wrapper


class MockLogger:
    def __call__(self, message, level="info"):
        print("{}: {}".format(level.upper(), message))


class Metrics:
    __version__ = "2.1.2"

    def __init__(
        self,
        logger,
        processor_name: str,
        uri_type: str,
        uri: str,
        bearer_token: str,
        batch_size: int = 20,
        flush_interval: int = 10,
        max_cache_size: int = 250,
        failure_drop_message_threshold: int = 100,
        failure_drop_time_threshold: int = 300,
        request_timeout: float = 10.0,
    ) -> None:
        self.logger = logger
        self.processor_name = processor_name
        self.uri_type = uri_type.lower()
        self.uri = uri
        self.bearer_token = bearer_token

        self.batch_size = batch_size
        self.flush_interval = flush_interval
        self.max_cache_size = max_cache_size
        self.failure_drop_message_threshold = failure_drop_message_threshold
        self.failure_drop_time_threshold = failure_drop_time_threshold
        self.request_timeout = request_timeout

        self.valid_uri_types = ["aws_lambda", "aws_api_gateway", "self-hosted"]

        self._metric_cache = []
        self._flush_scheduled = False
        self._sending = False

        self._failed_message_count = 0
        self._first_failure_time = None

        self._validate_settings()

    def _validate_settings(self) -> None:
        if self.uri_type not in self.valid_uri_types:
            raise ValueError(
                "URI Type must be one of: {}".format(str(self.valid_uri_types))
            )

        if self.uri_type == "aws_api_gateway":
            raise NotImplementedError(
                "aws_api_gateway is not implemented yet for AWS Serverless Metrics"
            )

        if not self.bearer_token:
            raise ValueError("Bearer token can not be empty")

        if self.batch_size < 1:
            raise ValueError("batch_size must be at least 1")

        if self.flush_interval <= 0:
            raise ValueError("flush_interval must be greater than 0")

        if self.max_cache_size < self.batch_size:
            raise ValueError(
                "max_cache_size must be greater than or equal to batch_size"
            )

        if self.failure_drop_message_threshold < 1:
            raise ValueError("failure_drop_message_threshold must be at least 1")

        if self.failure_drop_time_threshold <= 0:
            raise ValueError("failure_drop_time_threshold must be greater than 0")

        if self.request_timeout <= 0:
            raise ValueError("request_timeout must be greater than 0")

        self.logger("Metrics Settings validated successfully", "info")

    def _can_drop_oldest_metrics(self) -> bool:
        if self._failed_message_count >= self.failure_drop_message_threshold:
            return True

        if self._first_failure_time is not None:
            failure_duration = time.time() - self._first_failure_time

            if failure_duration >= self.failure_drop_time_threshold:
                return True

        return False

    def _trim_cache_if_allowed(self) -> None:
        if len(self._metric_cache) <= self.max_cache_size:
            return

        if not self._can_drop_oldest_metrics():
            self.logger(
                "Metric cache size is {}, max is {}, but dropping is not allowed yet.".format(
                    len(self._metric_cache),
                    self.max_cache_size,
                ),
                "warning",
            )
            return

        drop_count = len(self._metric_cache) - self.max_cache_size

        self.logger(
            "Dropping {} oldest cached metrics after send failures. Failed message count: {}".format(
                drop_count,
                self._failed_message_count,
            ),
            "warning",
        )

        self._metric_cache = self._metric_cache[drop_count:]

    def _cache_metric(self, action: str, metric: str) -> None:
        self._metric_cache.append(
            {
                "clientname": self.processor_name,
                "timestamp": datetime.now(timezone.utc).isoformat(),  # noqa: UP017
                "metric": metric,
                "action": action,
            }
        )

        self._trim_cache_if_allowed()

        if len(self._metric_cache) >= self.batch_size:
            self.flush()
        else:
            self._schedule_flush()

    def _schedule_flush(self) -> None:
        if self._flush_scheduled:
            return

        self._flush_scheduled = True

        @Wait(self.flush_interval)
        def _delayed_flush():
            self._flush_scheduled = False
            self.flush()

    def flush(self) -> None:
        if self._sending:
            self._schedule_flush()
            return

        if not self._metric_cache:
            return

        batch = self._metric_cache[: self.batch_size]
        self._metric_cache = self._metric_cache[self.batch_size :]
        self._sending = True

        try:
            self._send_batch(batch)
        except Exception as e:
            self.logger(e, "error")
            self._sending = False
            self._metric_cache = batch + self._metric_cache
            self._schedule_flush()

    def _send_batch(self, messages) -> None:
        def _send_batch_inner(messages):
            data = json.dumps({"messages": messages}).encode("utf-8")

            req = urllib.request.Request(
                self.uri,
                data=data,
                headers={
                    "Content-Type": "application/json",
                    "Authorization": "Bearer {}".format(self.bearer_token),
                },
                method="POST",
            )

            try:
                with urllib.request.urlopen(
                    req,
                    timeout=self.request_timeout,
                ) as response:
                    response_body = response.read().decode("utf-8", "replace")
                    # self.logger(
                    #     "Server response status: {}".format(response.status),
                    #     "info",
                    # )
                    # if response_body:
                    #     self.logger(response_body, "info")

                    ack_ok, ack_error = self._validate_server_ack(
                        response_body,
                        len(messages),
                    )

                    if ack_ok:
                        self._failed_message_count = 0
                        self._first_failure_time = None
                    else:
                        self.logger(
                            "Server acknowledgement validation failed: {}. Will retry {} metrics.".format(
                                ack_error,
                                len(messages),
                            ),
                            "error",
                        )
                        self._record_send_failure(messages)

            except urllib.error.HTTPError as e:
                if e.code == 429:
                    self.logger(
                        "Server rate limited request (429). Will retry {} metrics.".format(
                            len(messages)
                        ),
                        "warning",
                    )
                    self._record_send_failure(messages)

                elif 400 <= e.code < 500:
                    self.logger(
                        "Permanent client error {}. Dropping {} metrics.".format(
                            e.code,
                            len(messages),
                        ),
                        "error",
                    )

                    try:
                        self.logger(e.read(), "error")
                    except Exception:
                        pass

                else:
                    self.logger(
                        "Server error {}. Will retry {} metrics.".format(
                            e.code,
                            len(messages),
                        ),
                        "error",
                    )
                    self._record_send_failure(messages)

            except urllib.error.URLError as e:
                self.logger(e, "error")
                self._record_send_failure(messages)

            except Exception as e:
                self.logger(e, "error")
                self._record_send_failure(messages)

            finally:
                self._sending = False

                if self._metric_cache:
                    self._schedule_flush()

        @Wait(0)
        def _multi_thread_metric_send():
            _send_batch_inner(messages)

    def _validate_server_ack(self, response_body, expected_count):
        if not response_body:
            return False, "empty response body"

        try:
            ack = json.loads(response_body)
        except ValueError as error:
            return False, "invalid JSON response: {}".format(error)

        if not isinstance(ack, dict):
            return False, "response JSON was not an object"

        if ack.get("ok") is not True:
            return False, "response ok was not true"

        count = ack.get("count")
        if isinstance(count, bool) or not isinstance(count, int):
            return False, "response count was not an integer"

        if count != expected_count:
            return False, "response count {} did not match sent count {}".format(
                count,
                expected_count,
            )

        return True, ""

    def _record_send_failure(self, messages) -> None:
        if self._first_failure_time is None:
            self._first_failure_time = time.time()

        self._failed_message_count += len(messages)

        # Requeue failed messages at the front so they are retried before newer metrics.
        self._metric_cache = messages + self._metric_cache

        self._trim_cache_if_allowed()

    def start(self, metric_name: str) -> None:
        """Record a 'Started' event for a metric.

        args:
            metric_name (str): The name of the metric to record.
        """
        self._cache_metric("Started", metric_name)

    def stop(self, metric_name: str) -> None:
        """Record a 'Stopped' event for a metric.

        args:
            metric_name (str): The name of the metric to record.
        """
        self._cache_metric("Stopped", metric_name)

    def trace(self, metric_name: str) -> None:
        """Record a non-timebound event, such as a button press.

        args:
            metric_name (str): The name of the metric to record.
        """
        self._cache_metric("Trace", metric_name)

    def connected(self, metric_name: str) -> None:
        """Record a 'Connected' event for a metric.

        args:
            metric_name (str): The name of the metric to record.
        """
        self._cache_metric("Connected", metric_name)

    def disconnected(self, metric_name: str) -> None:
        """Record a 'Disconnected' event for a metric.

        args:
            metric_name (str): The name of the metric to record.
        """
        self._cache_metric("Disconnected", metric_name)

    def custom(self, action: str, metric_name: str) -> None:
        """Record a custom action.

        args:
            action (str): The custom action to record.
            metric_name (str): The name of the metric to record.
        """
        self._cache_metric(action, metric_name)


if __name__ == "__main__":
    # Simple test that runs on a workstation
    uri = "<your_uri>"
    bearer = "<your_bearer_token>"

    logger = MockLogger()

    test = Metrics(
        logger,
        "test_pc",
        "aws_lambda",  # or 'self-hosted'
        uri,
        bearer,
        flush_interval=10,
    )
    print("Testing Metrics...")
    print(
        "This will take {} seconds due to flush interval...".format(
            test.flush_interval + 5
        )
    )

    test.trace("Hello World!")
    test.start("TestStart")
    test.stop("TestStop")
    test.custom("CustomAction", "TestCustom")

    # Keep workstation script alive long enough for delayed flush.
    time.sleep(test.flush_interval + 5)
    print("Test complete.")
